VSMac support

nosami opened the pull request:
Nothing really works yet - but at least it is hosted :) I'll work on this some more in the coming days but thought you might like to see some progress.

<img width="1139" alt="image" src="https://user-images.githubusercontent.com/667194/65395557-5f94db00-dd94-11e9-9f3f-67096d863363.png">

**Edit**: Nearly everything is working. Build instructions are [here](https://github.com/VsVim/VsVim/pull/2733#issuecomment-538998555)

nosami commented:
First snag - 

@richsladkey You mention [here](https://github.com/VsVim/VsVim/issues/2655#issuecomment-506794388) 

> The provided extension point in Visual Studio that VsVim is currently using is explicitly WPF-based (KeyProcessor)

The provided extension point in VSMac is explicitly Cocoa based (See the NSEvent and ICocoaTextView references) [here](https://github.com/microsoft/vs-editor-api/commit/f3401b8c45ebe24867d6c8d15ccd539e7433cd4c#diff-582be9d8cf0cdc6bca19d67e07165bf3) . VS and VSMac both can use IKeyProcessorProvider but the interfaces are different on different platforms :(

nosami commented on Src/VimMac/KeyProcessorProvider.cs:
```diff
@@ -0,0 +1,144 @@
+    [ContentType(VimConstants.AnyContentType)]
+    [TextViewRole(PredefinedTextViewRoles.Editable)]
+    [Name("Key Processor Provider")]
+    internal class KeyProcessorProvider :
```
I couldn't easily use IKeyProcessorProvider so trying out a simpler IChainedCommandHandler<TypeCharCommandArgs>. I don't think this will allow me to check for all key combinations, but it gets me a little further for now.

nosami commented:
![vsvim](https://user-images.githubusercontent.com/667194/65452989-2ca52280-de3a-11e9-9311-88e460e4d88f.gif)

Some more stuff is working... I have some keyboard support and it looks like many things are just working. All normal mode commands seem to be working and I can make visual selections.

jaredpar commented:
This is exciting to see!

ricksladkey commented:
Woohoo, this is really happening! I am so glad, @nosami, that you're courageous enough to tackle this. I'll try to follow along and help wherever I can.

nosami commented:
I ported big chunks of the key processing code. Control keys are working now except for `<C-v>` for some reason.

nosami commented:
This is actually working fairly well now. Insert mode works now (didn't previously), there is a readonly status bar 
<img width="723" alt="image" src="https://user-images.githubusercontent.com/667194/65791119-f678e300-e158-11e9-8255-53f1eda3ceb2.png">

`=` is working

Much of the tab / window code is internal in VSMac (See here for an example of what I had to do in XSVim https://github.com/nosami/XSVim/blob/9e0f99ffcc7c8239d80fef6a75d59160c63adac9/XSVim/WindowManagement.fs#L26). 
Notice the use of a custom `?` operator which is used for reflection into internal types. I'm tempted to write this part of the code in F# so that I can do the same thing.
Maybe it's possible to use dynamic in C# to do the same?

Anyway, this is on pause until next week. Main things left to solve are figuring out why the ctrl keys aren't getting handled correctly and emulation of the block cursor.

nosami commented:
I've hit a snag that I'm not sure how to work around yet due to differences in how events are processed on Mac vs Windows.

With `KeyProcessor`, I am able to handle keyboard repeat, but not prevent keys from being entered into the editor. `NSEvent` is asynchronous and observable only. There isn't a `Handled` flag like there is on the Windows version of `KeyProcessor`

With `TypedCharCommand` I can prevent propagation, but I can't handle keyboard repeat. (Holding down the `c` key on a Mac pops up a native window that allows you to enter `č` for example)

It might be necessary for me to change code inside the editor itself, I'm not sure yet.

nosami commented:
The VSMac editor team are going to make a change to the editor to allow events to stop propagating from `KeyProcessor` to align better with how Windows handles keys. My guess is that this change wouldn't be released until 8.4.

For now, I have a bit of a hack that works most of the time.  I am using `KeyProcessor` to handle vim actions and `TypeCharCommandArgs` to handle actual input or keys that the IDE needs to handle outside of Vim. I am setting a property on the `TextView` from `KeyProcessor` which specifies whether or not the key was handled by Vim.

ssameng commented:
Just wanted to say I'm super grateful for you taking a stab at this! Thank you :)

nosami commented:
If anyone wants to run this Extension on VSMac (it's actually fairly usable already, I'm using it every day), then clone my fork and running the following commands should be enough.

```
msbuild /t:Restore
cd Src/VimMac
msbuild /p:InstallAddin=True
```

After running those commands, ensure that the Addin has been loaded into VSMac and is enabled.

 If you want to work on the code, then you will also need to install the AddinMaker extension

<img width="828" alt="image" src="https://user-images.githubusercontent.com/667194/66313143-c45e4280-e909-11e9-8356-91ff8e068e62.png">

This extension adds some IDE support. It takes care of locating VSMac references that are located under `/Applications/Visual Studio.app` and also makes it easy to run any changes (Start Debugging on the VimMac project should launch a new instance of VSMac with any changes to the Extension applied.

<img width="423" alt="image" src="https://user-images.githubusercontent.com/667194/66313484-5bc39580-e90a-11e9-8091-bb75e00f94ae.png">

I've been testing this with the latest stable (8.3.1.18 at the time of writing)

nosami commented:
On Windows, when intellisense is active in insert mode, does pressing the escape key just close the intellisense popup and remain in Insert mode or does it close the intellisense popup and switch to Normal mode?

I am leaning towards switching to Normal mode, but I suppose the behavior should be the same on Windows and Mac.

ssameng commented:
Thanks for posting instructions--I'll be trying out the Mac VsVim addin today.
Let me know if there's any testing to be done and I'm happy to help.

> On Windows, when intellisense is active in insert mode, does pressing the escape key just close the intellisense popup and remain in Insert mode or does it close the intellisense popup and switch to Normal mode?

Just checked this--it closes the intellisense popup and switches back to Normal mode.

nosami commented:
@ssameng thanks for the offer!

I'm especially interested in what works on VSWin but doesn't on VSMac if you are familiar with VSVim on Windows already :)

ssameng commented:
Just installed it--it didn't work right away (even after restarting VSMac and making sure it was enabled)
a computer restart and clearing the solution worked for me to enable it.

Showed it to my friend and we put it through a little test run, comparing it to default VsVim Windows.

We thought it was pretty solid. Great work!

A minor oddity: pressing backslash in Normal mode inserts a backslash and switches to Insert mode. Same with Shift-K and Shift-Q.

nosami commented:
@ssameng 

> A minor oddity: pressing backslash in Normal mode inserts a backslash and switches to Insert mode. Same with Shift-K and Shift-Q.

confirmed. weird!

nosami commented:
```vim
set incsearch
set hlsearch
inoremap jk <esc>
inoremap kj <esc>
inoremap jj <esc>
```

works :O

nosami commented:
> A minor oddity: pressing backslash in Normal mode inserts a backslash and switches to Insert mode. Same with Shift-K and Shift-Q.

Fixed
> Just checked this--it closes the intellisense popup and switches back to Normal mode.

Fixed

ssameng commented:
Great work so far. I've been more productive because of your plugin!

A difference I've noted between your plugin and VsVim windows.
On your latest commit as of this post.

Given the line:
`this.object.property`

In Normal Mode, put cursor on the `t` of `object`
Enter visual mode with `v`
Press `b` to select all of `object`
Press `x` to delete

Result on Windows:
`this..property`

Result on Mac:
`this.t.property`

Similar problem with starting at the beginning of a word and using `vwx`
Result on Windows:
`this.property`

Result on Mac:
`this..property`

nosami commented:
> Result on Mac:
this.t.property

That's odd.... when I try that I get `this.property`. I'm not sure what could be causing the difference.

Therzok commented on Src/VimMac/ClipboardDevice.cs:
```diff
@@ -0,0 +1,13 @@
+    [Export(typeof(IClipboardDevice))]
+    internal sealed class ClipboardDevice : IClipboardDevice
+    {
+        public bool ReportErrors { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
```
```suggestion
        public bool ReportErrors { get; set; }
```

Therzok commented on Src/VimMac/OsxClipboard.cs:
```diff
@@ -0,0 +1,69 @@
+
+    static OsxClipboard()
+    {
+        utfTextType = objc_msgSend(objc_msgSend(nsString, allocRegister), initWithUtf8Register, "public.utf8-plain-text");
```
Can we not add a dependency to Xamarin.Mac? This makes the assembly unable to benefit from static registrar codegen.

nosami replied:
I don't know how to handle the OSX Clipboard any other way

Therzok replied:
Are there no bindings to it in XamMac?

Therzok replied:
I think we can do something slightly better here. We could just get the IntPtr handles via these selectors. Then do `Runtime.GetNSObject<NSPasteBoard> (handle);` for them and try to use as much managed bindings that xammac offers.

For the general pasteboard, there's `NSPasteboard.GeneralPasteboard`.

Therzok replied:
I see that initWithUTF8String is not bound for NSPasteBoard, so we need to use objc_msgSend.

Therzok replied:
OHHH, this one is simple:

```csharp
NSString utfTextType;

utfTextType = new NSString ("public.utf8-plain-text");
```

Therzok commented on Src/VimMac/OsxClipboard.cs:
```diff
@@ -0,0 +1,69 @@
+    {
+        var ptr = objc_msgSend(generalPasteboard, stringForTypeRegister, nsStringPboardType);
+        var charArray = objc_msgSend(ptr, utf8Register);
+        return Marshal.PtrToStringAnsi(charArray);
```
```suggestion
        return Marshal.PtrToStringUTF8(charArray);
```

Therzok commented on Src/VimMac/OsxClipboard.cs:
```diff
@@ -0,0 +1,69 @@
+        {
+            if (str != default)
+            {
+                objc_msgSend(str, sel_registerName("release"));
```
Cache this, selector lookup is expensive.

Therzok commented on Src/VimMac/DefaultKeyProcessorProvider.cs:
```diff
@@ -0,0 +1,37 @@
+    [ContentType(VimConstants.ContentType)]
+    [TextViewRole(PredefinedTextViewRoles.Interactive)]
+    [Name(VimConstants.MainKeyProcessorName)]
+    [Order(After = "CtrlClickGoToDefKeyProcessor")]
```
```suggestion
    [Order(After = nameof(CtrlClickGoToDefKeyProcessor))]
```

Therzok commented on Src/VimMac/StatusBar.cs:
```diff
@@ -0,0 +1,168 @@
+                        : string.Format("-- ({0}) SELECT LINE --", oneTimeArgument);
+                    break;
+                case ModeKind.ExternalEdit:
+                    status = "External edit detected (hit &lt;Esc&gt; to return to previous mode, &lt;C-c&gt; to cancel external edit)";
```
Should these be translated?

nosami replied:
They will be at some point once I/We have decided how to best handle shared translations between Wpf/Mac

nosami replied:
This whole class might disappear at some point

Therzok commented on Src/VimMac/VimHost.cs:
```diff
@@ -0,0 +1,593 @@
+        {
+            using (var stream = typeof(VimCocoaHost).Assembly.GetManifestResourceStream("Vim.Mac.Resources.beep.wav"))
+            {
+                NSData data = NSData.FromStream(stream);
```
Wrapping this in a `using` will prevent finalizer from running.

Therzok commented on Src/VimMac/WindowManagement.cs:
```diff
@@ -0,0 +1,81 @@
+
+    internal static class WindowManagement
+    {
+        static BindingFlags instanceFlags = BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public;
```
```suggestion
        const BindingFlags instanceFlags = BindingFlags.Instance | BindingFlags.NonPublic | BindingFlags.Public;
```

Therzok commented on Src/VimMac/OsxClipboard.cs:
```diff
@@ -0,0 +1,69 @@
+    static OsxClipboard()
+    {
+        utfTextType = objc_msgSend(objc_msgSend(nsString, allocRegister), initWithUtf8Register, "public.utf8-plain-text");
+        nsStringPboardType = objc_msgSend(objc_msgSend(nsString, allocRegister), initWithUtf8Register, "NSStringPboardType");
```
This can just be `new NSString("NSStringPboardType")`

Therzok commented on Src/VimMac/OsxClipboard.cs:
```diff
@@ -0,0 +1,69 @@
+        utfTextType = objc_msgSend(objc_msgSend(nsString, allocRegister), initWithUtf8Register, "public.utf8-plain-text");
+        nsStringPboardType = objc_msgSend(objc_msgSend(nsString, allocRegister), initWithUtf8Register, "NSStringPboardType");
+
+        generalPasteboard = objc_msgSend(nsPasteboard, generalPasteboardRegister);
```
`NSPasteboard.GeneralPasteboard`

Therzok commented on Src/VimMac/OsxClipboard.cs:
```diff
@@ -0,0 +1,69 @@
+
+    public static string GetText()
+    {
+        var ptr = objc_msgSend(generalPasteboard, stringForTypeRegister, nsStringPboardType);
```
Given we have an `NSString` statically constructed, we could just do:

```csharp
var ptr = objc_msgSend(generalPasteboard.Handle, stringForTypeRegister, nsStringPboardType.Handle);
return NSString.FromHandle (ptr);
```

Therzok commented on Src/VimMac/OsxClipboard.cs:
```diff
@@ -0,0 +1,69 @@
+        {
+            str = objc_msgSend(objc_msgSend(nsString, allocRegister), initWithUtf8Register, text);
+            objc_msgSend(generalPasteboard, clearContentsRegister);
+            objc_msgSend(generalPasteboard, setStringRegister, str, utfTextType);
```
ClearContents can be migrated to xammac bindings.

```csharp
var str = default;
try {
     generalPasteboard.ClearContents();
     str = NSString.CreateNative (text);
     objc_msgSend(generalPasteboard.Handle, setStringRegister, str, utfTextType.Handle);
} finally {
     NSString.ReleaseNative (str);
}
```

nosami commented:
@Therzok I've given you collaborator access :)

aaronlab commented:
I'm noticing that XAML files the editor is going bonkers for some reason.  Is there a way to shut down the mappings based on file type?

nosami commented:
@aaronlab I have started a branch that builds against VSMac 8.4 https://github.com/nosami/VsVim/commits/vsmac-8.4

The XAML bug is gone there. Thanks for the report!

I'll merge that branch to this one once 8.4 is made public.

akamud commented:
I've tried porting this before with no luck, probably because I don't know F# at all, but now that you have done all the heavy work, is there any way we can help you port the code? Is there anything specially broken atm the we might take a look?

nosami commented:
@akamud Please shout out if you do decide to take something on. Contact me either here or on [XSVim gitter](https://gitter.im/XSVim/Lobby) . Also, feel free to shout out for any other reason. Cheers!

nosami commented:
Added a Vim output console pad. Maybe we could use an editor for this in the future so that Vim works there too?

<img width="912" alt="Screenshot 2019-10-19 at 14 27 37" src="https://user-images.githubusercontent.com/667194/67147870-d7b9c800-f290-11e9-83b9-b9c8dbdd3635.png">

Therzok commented on Src/VimMac/VimConsoleFactoryService.cs:
```diff
@@ -0,0 +1,27 @@
+        [ImportingConstructor]
+        internal VimConsoleFactoryService()
+        {
+            var monitors = (IdeProgressMonitorManager)Runtime.GetService<ProgressMonitorManager>().Result;
```
Should we use PeekService here?

nosami replied:
I changed this code already :)

nosami commented:
> A few questions:
> 
> Do we want to reuse a single output pad for all the buffers? 
Yes. Although Vim behaviour is a little different. It displays the message and then disappears when you press Enter. This is just a quick win solution until I think of a better way.

<img width="581" alt="image" src="https://user-images.githubusercontent.com/667194/67149939-591d5480-f2a9-11e9-8395-94b19b6a5910.png">


> Maybe annotate about which buffer the message comes from?
Vim doesn't do this

nosami commented:
It's now possible to run shell commands like `:! sort` on a text selection or `:r ! ls -la`.

If you launch from the terminal, then VSMac will inherit your shells $PATH environment variable and the commands will just work.

If you launch VSMac from the dock or application launcher then run the following command `launchctl setenv PATH $PATH` to mirror your PATH environment variable to OSX apps.

nosami commented:
`:vimgrep` is now working. Will need `$PATH` support as mentioned in the last message.

Forms `:vimgrep Vim ~/src/VsVim` and `:vimgrep Vim ~/src/VsVim/**/*.fs` are both working

nosami commented:
VSMac 8.4 is available on the Internal Preview channel. If you are able to, please use this version as there have been some bug fixes in the editor to better support VsVim.

nosami commented:
Added 2 helper commands to open VSMac pads to make it possible to open pads via `~/.vimrc` or `~/.vsvimrc` keybindings.

The commands are `VsVim.ShowPadById` and `VsVim.ShowPadByTitle`.

Example usage 

```
nmap gsol :vsc VsVim.ShowPadByTitle Solution<CR>
```

Therzok commented on Src/VimMac/AlternateKeyUtil.cs:
```diff
@@ -0,0 +1,256 @@
+        internal static NSEventModifierMask ConvertToModifierKeys(VimKeyModifiers keys)
+        {
+            NSEventModifierMask res = 0;
+            if (0 != (keys & VimKeyModifiers.Shift))
```
Consistency: Use HasFlags as above. Mono/.netcore have runtime instrincs for them, so both versions will generate the same code

nosami commented:
@ssameng Do you still have the issue that you mentioned earlier? I haven't seen it yet.

> Great work so far. I've been more productive because of your plugin!
> 
> A difference I've noted between your plugin and VsVim windows.
> On your latest commit as of this post.
> 
> Given the line:
> `this.object.property`
> 
> In Normal Mode, put cursor on the `t` of `object`
> Enter visual mode with `v`
> Press `b` to select all of `object`
> Press `x` to delete
> 
> Result on Windows:
> `this..property`
> 
> Result on Mac:
> `this.t.property`
> 
> Similar problem with starting at the beginning of a word and using `vwx`
> Result on Windows:
> `this.property`
> 
> Result on Mac:
> `this..property`

ssameng commented:
Been using VsVim Mac quite extensively. Of course, I am aware it is in progress--but I've run into a two main bugs that unfortunately keep me from using it fulltime

1. Backspace key doesn't work in insert mode past the current "buffer".
Reproduction: In a file, enter insert mode. Press backspace. The chime plays, and delete doesn't run. It works to delete any character only if you typed that character during insert mode.

2. Refactoring doesn't work.
Reproduction: In a C# file, given `var x = 1;`, right click on `x` and select Rename. VS is now locked up and insert mode cannot be entered. Closing the tab and reloading the file fixes this.

Let me know if you need any more information!

nosami commented:
> Backspace key doesn't work in insert mode past the current "buffer".
Reproduction: In a file, enter insert mode. Press backspace. The chime plays, and delete doesn't run. It works to delete any character only if you typed that character during insert mode.

How did I not notice this? I must not use the backspace key often :)

> Refactoring doesn't work.
Reproduction: In a C# file, given var x = 1;, right click on x and select Rename. VS is now locked up and insert mode cannot be entered. Closing the tab and reloading the file fixes this.

Yeah, I've seen this too but haven't investigated yet.

@ssameng Thanks for the report!

nosami commented:
@ssameng Inline rename is fixed

ssameng commented:
> If the chime plays, does that mean you are at the start of the file? Does this happen in any file?

I'm not sure if it means I'm at the start of the file--I do get the same metallic chime when I try navigating upwards or backwards while at the start of the file.
It happens in every C# file so far.
The backspace key works correctly during Normal mode, and does the chime when I'm at beginning of file.

nosami commented:
@ssameng Are you using VSMac 8.3 or 8.4? I can't repro on 8.4

Therzok commented on Src/VimMac/CommandKeyBinding.cs:
```diff
@@ -0,0 +1,124 @@
+
+        public override bool Equals(object obj)
+        {
+            if (obj is CommandId)
```
Nit:
```csharp
return obj is CommandId cmd && Equals (cmd);
```

nosami replied:
This was just copied verbatim from a file that was probably written years ago.

I hope that at some point, once we figure out which code is common between VSMac and VS Windows, that we can come up with a plan for sharing the code instead of copying.

jaredpar replied:
> I hope that at some point, once we figure out which code is common between VSMac and VS Windows, that we can come up with a plan for sharing the code instead of copying.

Feel the same way. I think once we get the VSMac code merged and a bit of experience with developers using it, it will become fairly clear what can / should be shared with the Windows version. At that point we can start the work of defining that layer and moving VS Windows and VS Mac to depend on it.

ssameng commented:
> @ssameng Are you using VSMac 8.3 or 8.4? I can't repro on 8.4

Ah, good point. I'm on the latest stable, 8.3.6
I'll have a go at 8.4

ssameng commented:
> > @ssameng Are you using VSMac 8.3 or 8.4? I can't repro on 8.4
> 
> Ah, good point. I'm on the latest stable, 8.3.6
> I'll have a go at 8.4


Just removed the plugin, updated to 8.4 preview and built branch `vsmac-8.4` and installed the plugin--Same problem with backspace in Insert mode.
Any other things that might be different in my environment?
I'm still on macOS 10.14.6

nosami commented:
> Any other things that might be different in my environment?
I'm still on macOS 10.14.6

I have no idea what could be causing the issue. What does `past the current "buffer".` mean? Can you give me some repro instructions? Does backspace in insert mode _ever_ work for you?

ssameng commented:
> I have no idea what could be causing the issue. What does `past the current "buffer".` mean? Can you give me some repro instructions? Does backspace in insert mode _ever_ work for you?

Ah, sorry--Given the line
`var x = 0;`
1. In Normal Mode, backspace works to move cursor back.
2. Move cursor on the `;`
3. Enter Insert mode with `i`
4. Press backspace. Chime plays, no backspace happens
5. Type `123`
6. Line is now
`var x = 0123;`
7. Press backspace. It works now, and can erase the `123` just typed.
8. Chime plays once the cursor hits the `0` and no more backspaces are possible

nosami commented:
@ssameng 

![2019-11-03 21 37 15](https://user-images.githubusercontent.com/667194/68092362-2dda5e00-fe82-11e9-80e5-a1201c78d11b.gif)

Works on my machine :) No idea what could be going wrong for you. Interested to know if anyone else is seeing this.

ssameng commented:
Hmm, super weird!
Perhaps something in the preferences?

Can you upload your Library/Preferences/VisualStudio/8.0 ?

Attached mine here.
[8.0.zip](https://github.com/VsVim/VsVim/files/3802200/8.0.zip)

nosami commented:
@ssameng Is it possible for you to debug this? It might be the only way to see what is going on. I would be interested in seeing a stacktrace from the [Beep method](https://github.com/nosami/VsVim/blob/bc1191585e1087a6f585692aae16f25e4807ca29/Src/VimMac/VimHost.cs#L104) if you're able to.

ssameng commented:
I'm happy to! I'm unsure of how to though--I guess build with a breakpoint, and then run the plugin while in debugging?

Could you write up a quick instructional?

nosami commented:
@ssameng Yeah... it's pretty much just what you said:)

Just need to make sure that AddinMaker is installed and then set a breakpoint on the Beep method, start debugging the VimMac project to launch a new instance of VSMac and then wait for the breakpoint to hit when you press Backspace. Feel free to shout on https://gitter.im/XSVim/Lobby if you get stuck. Cheers!



>  If you want to work on the code, then you will also need to install the AddinMaker extension

<img width="828" alt="image" src="https://user-images.githubusercontent.com/667194/66313143-c45e4280-e909-11e9-8356-91ff8e068e62.png">

> This extension adds some IDE support. It takes care of locating VSMac references that are located under `/Applications/Visual Studio.app` and also makes it easy to run any changes (Start Debugging on the VimMac project should launch a new instance of VSMac with any changes to the Extension applied.

<img width="423" alt="image" src="https://user-images.githubusercontent.com/667194/66313484-5bc39580-e90a-11e9-8091-bb75e00f94ae.png">

ssameng commented:
Thanks for the tips! I've successfully done it--looks like the backspacing command does believe there's no characters left.

<img width="2115" alt="Screen Shot 2019-11-03 at 5 35 43 PM" src="https://user-images.githubusercontent.com/5396947/68093135-610bf580-fe60-11e9-9dbd-f52246474276.png">

nosami commented:
So. I should have remembered that backspace is broken in Vim by default too. You need 
`set backspace=indent,eol,start` to make it work like you expect.

If anyone is interested, I have the following in my `~/.vsvimrc` right now

```vim
set incsearch
set hlsearch
let mapleader=" "
set cursorline
set clipboard=unnamed
set expandtab
set ignorecase
set smartcase
set autoindent
set backspace=indent,eol,start
set tabstop=4
set shiftwidth=4
nnoremap <C-p> :vsc MonoDevelop.Ide.Commands.SearchCommands.GotoFile <CR>
nnoremap gu :vsc MonoDevelop.Refactoring.RefactoryCommands.FindReferences <CR>
nnoremap gh :vsc MonoDevelop.Ide.Commands.TextEditorCommands.ShowQuickInfo <CR>
nnoremap gb :vsc MonoDevelop.RefactoryCommands.NavigationCommands.FindBaseSymbols <CR>
nnoremap <C-w>o :vsc MonoDevelop.Ide.Commands.FileTabCommands.CloseAllButThis <CR>
" I prefer default VSMac behaviour of navigating between buffers
" when navigating backwards and forwards
nnoremap <C-o> :vsc MonoDevelop.Ide.Commands.NavigationCommands.NavigateBack <CR>
nnoremap <C-i> :vsc MonoDevelop.Ide.Commands.NavigationCommands.NavigateForward <CR>
" Keys copied from SpaceMacs
nnoremap <Space>p :vsc MonoDevelop.Ide.Commands.SearchCommands.GotoFile <CR>

nmap gs :vsc VsVim.ShowPadByTitle Solution<CR>
```

ricksladkey commented:
> @ssameng Inline rename is fixed

Wow, following along and I am super impressed with the progress. The challenge was daunting but now there are valid reasons to be optimistic!

nosami commented:
I just spent a while trying to figure out why the caret was moving when yanking. Turns out it was already fixed [here](https://github.com/VsVim/VsVim/issues/2735)

Nice to see that Windows and Mac has the same bugs :)

nosami commented:
@jaredpar @ricksladkey 
It would be nice to get the build green again and get a build for Mac working on CI.

How would you like to separate the Windows build from the Mac build?
Separate solution configuration, separate solution file or something else?

Thanks!

nosami commented:
I excluded VimMac.csproj from the Debug/Release configurations so that the Windows build should work as before.

I've added DebugMac/ReleaseMac configurations that contain the projects that are buildable on Mac.

I noticed when doing this that I can't build any of the Test projects due to references on WPF. It would be nice to get this fixed for VimCoreTests at least so that Mac users can run this suite locally.

nosami commented:
Word completion. That was easy :O

<img width="927" alt="image" src="https://user-images.githubusercontent.com/667194/68405889-ca4d7a80-0178-11ea-98b2-e5b918c2e509.png">

jaredpar commented:
@nosami 

What exactly is blocking the VimMac.csproj from building on Windows? The reference DLLs could be checked in under `References` as we do for other reference assemblies. Is it a later tool that is causing the breaks?

> I noticed when doing this that I can't build any of the Test projects due to references on WPF. It would be nice to get this fixed for VimCoreTests at least so that Mac users can run this suite locally.

Once we get this merged we can start pulling apart the WPF reference in VimCoreTests so that they can run from Mac. That will be a bit of work but I think it's doable.

nosami commented:
@jaredpar I haven't attempted to build it on Windows yet but my gut feeling is that it would be pretty difficult. 

There are some Xamarin.Mac references such as `AppKit`. This is needed for native macOS types such as `NSEvent` and `NSKey`

nosami commented:
I found this forum post about building Xamarin.Mac apps on Windows https://forums.xamarin.com/discussion/59538/build-mac-apps-on-windows . The VSMac reference libraries could be added into this repo for sure... it's just the Xamarin.Mac framework stuff that would be tricky. I'll have a think about it. I put the binlog here - https://github.com/nosami/nosami.github.io/blob/master/msbuild.binlog?raw=true

I was just thinking that I would want to take this out of draft mode soon and start the feedback process. I'm sure there will be lots of it! I'll just go through myself and tidy things up and then move this PR to Ready for review. There is still tons to do of course, but this PR is in danger of getting out of hand :)

nosami commented:
This PR is based on VSMac 8.3 but I want to stop working on that and just get 8.4 (still in preview) working going forward (will require installing 8.4 on CI to build against)

The 8.3 code has hacks in to work around the event propagation issue I mentioned and shouldn't really be used. I'll switch this PR to my 8.4 branch and see what breaks.

nosami commented:
OK.. this PR is now mean for consumption by VSMac 8.4 clients only. If you've been following along and are still using VSMac 8.3, then you'll want to check out my vsmac-8.3 branch.

robert-hardy commented:
I installed VsVim on my VS for Mac, version 8.3.5 (build 13). Install was from git, I followed the instructions posted earlier here.
Couldn't initially get it to work so I did a clean/close/restart and still no luck.
Well, perhaps it was something else: I now find it works when I edit a plain text file, but does not work if I am in a .fs file nor a .fsx file.
Indeed, I can have a .txt and a .fs open and see it working in the .txt but not working when I switch to the .fs. I switch back to the text file and all is working again.

nosami commented:
@robert-hardy 

There are currently 2 different editors in VSMac. `C#`, `.txt` and a few of the web based languages are using the new editor and the rest (including F#) use the old editor.

VsVim works with the new editor only. For the old editor, you can install this [extension](https://github.com/nosami/XSVim/releases/download/0.65.11.81/XSVim.XSVim_0.65.11.81.mpack) via `Visual Studio -> Extensions -> Install from file`. You can use both extensions side by side without any problems.

I will be working on making F# work in the new editor soon. Once that's done, you will be able to stop using the extra extension.

donquixote1984 commented:
love this.

otaviopace commented:
> If anyone wants to run this Extension on VSMac (it's actually fairly usable already, I'm using it every day), then the following commands should be enough. Please use VSMac 8.4 if possible - the editor contains some fixes needed for this extension.
> 
> #### For VSMac 8.3
> ```shell
> git clone --branch vsmac-8.3 https://github.com/nosami/VsVim.git
> cd VsVim
> msbuild /t:Restore
> cd Src/VimMac
> msbuild /p:InstallAddin=True
> ```
> 
> or
> 
> ```shell
> git clone --branch vsmac-8.3 https://github.com/nosami/VsVim.git && cd VsVim && msbuild /t:Restore && cd Src/VimMac && msbuild /p:InstallAddin=True
> ```
> 
> #### For VSMac 8.4
> ```shell
> git clone --branch vsmac https://github.com/nosami/VsVim.git
> cd VsVim
> msbuild /t:Restore
> cd Src/VimMac
> msbuild /p:InstallAddin=True
> ```
> 
> or
> 
> ```shell
> git clone --branch vsmac https://github.com/nosami/VsVim.git && cd VsVim && msbuild /t:Restore && cd Src/VimMac && msbuild /p:InstallAddin=True
> ```
> 
> This will build the VsVim extension against the current installation of VSMac.
> 
> After running those commands, ensure that the Addin has been loaded into VSMac and is enabled.
> 
> <img alt="image" width="815" src="https://user-images.githubusercontent.com/667194/66313771-d42a5680-e90a-11e9-9d00-cb3563ef4a9e.png">
> 
> If you want to work on the code, then you will also need to install the AddinMaker extension
> 
> <img alt="image" width="828" src="https://user-images.githubusercontent.com/667194/66313143-c45e4280-e909-11e9-8356-91ff8e068e62.png">
> 
> This extension adds some IDE support. It takes care of locating VSMac references that are located under `/Applications/Visual Studio.app` and also makes it easy to run any changes (Start Debugging on the VimMac project should launch a new instance of VSMac with any changes to the Extension applied.
> 
> <img alt="image" width="423" src="https://user-images.githubusercontent.com/667194/66313484-5bc39580-e90a-11e9-8091-bb75e00f94ae.png">
> 
> I've been testing this with the latest stable (8.3.1.18 at the time of writing). It _might_ work on other versions, but hasn't been tested.

At least on VSMac 8.3 it is missing another `msbuild /t:Restore` after `Src/VimMac`, at least on my Mac it errored without this. So the full set of commands would be this:

```
git clone --branch vsmac-8.3 https://github.com/nosami/VsVim.git
cd VsVim
msbuild /t:Restore
cd Src/VimMac
msbuild /t:Restore
msbuild /p:InstallAddin=True
```

Maybe the first `Restore` isn't necessary, I don't know.

abiosoft commented:
Thanks for this plugin, amazing work.
The only issue I'm having is refactors. I can't refactor unless the plugin is disabled.

I always get this message 
![Screenshot 2019-12-14 at 05 15 48](https://user-images.githubusercontent.com/240448/70843363-49e1ff80-1e31-11ea-85e4-0e0baf2635e4.png)

Thanks.

nosami commented:
@abiosoft Is that when running "Inline rename"? It looks like an error that I fixed a while ago. Did you build from the latest commit? Thanks

nosami commented:
@otaviopace Thanks for letting me know. The Restore command needed to be changed since I updated the configuration. 

I updated the build instructions.

nosami commented:
@abiosoft I changed the code so that VSVim ignores keys while in inline rename mode.  I think it works better now.

Please can you try the latest commit?  Thanks!

abiosoft commented:
@nosami thanks it works now.

Apparently, VS mac does fail to refactor sometimes. I previously took all failed refactors to be related to the plugin, but that is not the case. I discovered after doing multiple tests with and without the plugin.

abiosoft commented:
If it helps, I'm on 8.4 preview (8.4 build 2637)

nosami commented:
> Apparently, VS mac does fail to refactor sometimes. I previously took all failed refactors to be related to the plugin, but that is not the case. I discovered after doing multiple tests with and without the plugin.

If you can find a reproducible example for this, please share!

abiosoft commented:
One minor thing I noticed is deleting highlighted text ignores the last character.

![2019-12-17 15 44 57](https://user-images.githubusercontent.com/240448/71005758-8721e800-20e4-11ea-8673-db949e8bcc65.gif)

And I have 2 questions.
1. Do you have to uninstall before installing a newer version of the plugin or simply running
the build instructions above is enough and it'll be automatically updated.
2. Should we keep commenting on this thread for anything noticed or new issue should be opened instead.

Once again, thanks for this plugin.

nosami commented:
> One minor thing I noticed is deleting highlighted text ignores the last character.

Uh, yeah. Surprised I hadn't noticed that myself. I think the problem is that the caret is one too many places to the right when in visual mode (it should be at the last character of the selection but it is moved just past the selection).

Not sure what would be causing that as I think that is controlled by VimCore. I'll have to look into it.

> Do you have to uninstall before installing a newer version of the plugin or simply running
the build instructions above is enough and it'll be automatically updated.

Yeah, it should just work providing you have the latest code. `/p:InstallAddin` removes any previous version before installing the new version.

ricksladkey commented:
> > One minor thing I noticed is deleting highlighted text ignores the last character.
> 
> Uh, yeah. Surprised I hadn't noticed that myself. I think the problem is that the caret is one too many places to the right when in visual mode (it should be at the last character of the selection but it is moved just past the selection).

Be aware that there are or have been numerous bugs related to inclusive vs. exclusive selections and that any such bug **might** be a symptom of a problem unrelated to VSMac.

nosami commented:
> Be aware that there are or have been numerous bugs related to inclusive vs. exclusive selections and that any such bug might be a symptom of a problem unrelated to VSMac.

Ah, ok. I just assume that Vim.Core has no bugs! I guess I need to get a Windows machine set up so that I can do a comparison.

coolbluewater commented:
Thanks for the plugin! I've run into the refactoring/renaming issue discussed above. On my mac, with VsVim enabled, doing a rename refactoring does not work. The text is highlighted, and I can enter the new name, but when I press enter, the highlight disappears and the old name returns. No changes occur.

I have to disable the VsVim plugin and restart VSMac for renaming to start working again. (The rename itself is sometimes buggy and garbles text, but that I assume is a separate VSMac bug.)

I am using the latest Preview build of VSMac 8.4 (build 2653) and a build from the current sources of VsVim following the instructions above.

nosami commented:
@coolbluewater are you certain that you are building from the latest commit. What you describe sounds like what I fixed in https://github.com/VsVim/VsVim/pull/2733/commits/d0e31f2fdfa943252e7f7fcb243c8605f460d99b and I can't repro. Thanks!

coolbluewater commented:
I'm building against `93e869a`, is that correct?

nosami commented:
> I'm building against [nosami@93e869a](https://github.com/nosami/VsVim/commit/93e869a63e7812208be46d1d22884ed285651214), is that correct?

Yes, that's fine. I'm not sure what could be going wrong.

![rename](https://user-images.githubusercontent.com/667194/71560063-fdb3d380-2a5c-11ea-9af8-20e3b076c14e.gif)

It seems fine for me, but I did just observe that it got stuck once. I'll take another look at it when I get some more time. Thanks!

jcummings2 commented:
> It seems fine for me, but I did just observe that it got stuck once. I'll take another look at it when I get some more time. Thanks!

The fix in d0e31f2 makes renaming work almost all the time for me. It sometimes messes up when I have done a `git pull` or changed a file outside of VS. Restarting VS makes renaming work again.

jcummings2 commented:
> 2. Should we keep commenting on this thread for anything noticed or new issue should be opened instead.

I wondered the same thing about reporting any issues I encounter (none have been major).

Along those lines, I also wonder when this PR is going to be merged. I have been using it for a couple months now, so it certainly seems stable enough for widespread use.

jaredpar commented:
> Along those lines, I also wonder when this PR is going to be merged

My bad. I didn't realize this got moved out of Draft mode. I'd been holding off on a big review until that happened. Expected I would get some notification when it did. Will start my review today.

jaredpar commented on Src/VimMac/InlineRename/InlineRenameListenerFactory.cs:
```diff
@@ -0,0 +1,127 @@
+        {
+            return
+                contentType.IsCSharp() ||
+                //contentType.IsFSharp() ||
```
Just wanted to make sure this was deliberate as VSMac supports F#. It's not a blocker either way.

nosami replied:
Can't remember why I did this. I think it was because VSMac has no support for new editor and F#. But I am working on that :)

I think I should uncomment this.

jaredpar commented on Src/VimMac/IKeyUtil.cs:
```diff
@@ -0,0 +1,29 @@
+		/// </summary>
+		bool TryConvertSpecialToKeyInput(NSEvent theEvent, out KeyInput keyInput);
+	}
+}
```
Nit: tabs in this file instead of spaces

nosami replied:
Oops!

nosami commented:
> Incredibly excited about this work. I'm happy to merge whenever you are ready.

Good to hear. I know it's not ideal that the code can't be worked on easily from Windows but I'm happy to stick around and fix any Mac related issues.

nosami commented:
> I'm happy to merge whenever you are ready.

I'm ready for it to be merged now.

It's very usable already. I have more work still to do but I don't have time right now to do it. Nothing that I would consider a showstopper though. I'll get back to it as soon as I can.

Let me know if you want me to publish an Addin on the VSMac feed for 8.4 (8.3 behaves badly when editing .xaml files with VSVim). Unfortunately this step can't be automated, but I can talk you through it or just do it myself. Cheers!

jaredpar commented:
> Let me know if you want me to publish an Addin on the VSMac feed for 8.4

Yes please do so.

nosami commented:
> Yes please do so.

Done!

<img width="770" alt="Screenshot 2020-01-04 at 01 04 39" src="https://user-images.githubusercontent.com/667194/71757491-4240d400-2e8e-11ea-87ea-3e32117d19ba.png">

<img width="841" alt="Screenshot 2020-01-04 at 01 06 12" src="https://user-images.githubusercontent.com/667194/71757513-71efdc00-2e8e-11ea-898d-167d5d40cb17.png">

jaredpar commented:
That is absolutely wild to see!

altmnt commented:
For some reason when I switch to INSERT mode the Delete key is only working for the text I typed in that particular INSERT seession. E.g. if I just enter INSERT mode and hit Delete it wont delete previous character.
VS Community 8.4 (2653)
Vim plugin from Gallery - 2.8.0.0

nosami commented:
@altmnt The default settings need to be tweaked I think.

See here -  https://github.com/VsVim/VsVim/pull/2733#issuecomment-549187752

altmnt commented:
> @altmnt The default settings need to be tweaked I think.
> 
> See here - [#2733 (comment)](https://github.com/VsVim/VsVim/pull/2733#issuecomment-549187752)

That helped. Thanks.
